asllib/ASLTypingReference.mld

```diff
@@ -28,38 +28,113 @@ annotate a specific object, as follows.
 Typing of Expressions
 }
 
-[annotate_expr] specifies how to annotate an expression [e] in an
-environment [env]. 
-
-Typing of the expression [e] is <description> and one of the following applies:
+[annotate_expr] specifies how to annotate an expression [e] in an environment
+[env].  Formally, the result of annotating the expression [e] in [env] is
+[t,new_e] and one of the following applies:
 
 {3 TypingRule.Lit}
 
   {4 Rule}
+  All of the following applies:
+  - [e] is a Literal [v];
+  - [t] is the type annotation of [v];
+  - [new_e] is [e].
 
   {4 Example}
 
 {3 TypingRule.TypedExpr}
 
   {4 Rule}
-  [(e,_t)] denotes a typed expression and <description>
+  All of the following applies:
+  - [e] is a typed expression [(e',t')];
+  - [t'',e''] is the result of annotating [e'] in [env];
+  - One of the following applies:
+    - All of the following applies:
+      - [t''] is a structural subtype of [t'] in [env];
+      - [t''] is a domain subtype of [t'] in [env];
+      - [t] is [t'];
+      - [new_e] is [e'']. 
+    - All of the following applies:
+      - [t''] is a structural subtype of [t'] in [env];
+      - [t''] is not a domain subtype of [t'] in [env];
+      - an execution-time check that the expression evaluates to a value in the
+        domain of the required type is required.
+   - All of the following applies:
+     - [t''] is not a structural subtype of [t'] in [env];
+     - [t''] is not a domain subtype of [t'] in [env];
+     - a "ConflictingTypes" error is raised.
+ 
+  {4 Example}
+
+{3 TypingRule.ELocalVarConstant}
+
+  {4 Rule}
+  All of the following applies:
+  - [e] is a variable [x];
+  - [x] maps to a type [ty] in the [storage_types] of the local environment given by [env];
+  - [x] maps to a local constant [v];
+  - [t] is [ty];
+  - [new_e] is [E_Literal v].
```

Comment:
I don't think that we've been using the `AST.mli` constructors to deconstruct the AST (see for example at L73: "[e] is a variable [x]"), so I think we should say
```suggestion
  - [new_e] is the Literal [v].
```